NetKet/Machine/conv_layer.hpp

```diff
@@ -0,0 +1,435 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_CONVLAYER_HH
+#define NETKET_CONVLAYER_HH
+
+#include <time.h>
+#include <Eigen/Dense>
+#include <algorithm>
+#include <complex>
+#include <fstream>
+#include <random>
+#include <vector>
+#include "Lookup/lookup.hpp"
+#include "Utils/all_utils.hpp"
+#include "abstract_layer.hpp"
+
+namespace netket {
+/** Convolutional layer with spin 1/2 hidden units.
+ Important: In order for this to work correctly, VectorType and MatrixType must
+ be column major.
```

Comment:
Where does this restriction come from?

In any case we should make sure this restriction is enforced (otherwise we have a source of bugs). I'm not an expert on template programming, but the storage order [is a template parameter](https://eigen.tuxfamily.org/dox/group__TopicStorageOrders.html) of the `Eigen::Matrix` class, so I think it should be possible to check this at compile time.

Reply:
This restriction comes from the use of eigen::map.

Reply:
I added a check

Reply:
> I added a check

Thanks. I took the liberty to change it to a compile-time check in 67fe7e6 so that the compilation will abort with an error message if the conv layer is used with a row-major matrix.